code/messaging/src/main/java/com/adobe/marketing/mobile/MessagingPushTrackerActivity.java

```diff
@@ -0,0 +1,122 @@
+package com.adobe.marketing.mobile;
+import android.app.Activity;
+import android.app.NotificationManager;
+import android.app.TaskStackBuilder;
+import android.content.Context;
+import android.content.Intent;
+import android.net.Uri;
+import android.os.Bundle;
+import androidx.annotation.Nullable;
+
+import com.adobe.marketing.mobile.services.ServiceProvider;
+import com.adobe.marketing.mobile.util.StringUtils;
+
+public class MessagingPushTrackerActivity extends Activity {
+
+    @Override
+    protected void onCreate(final @Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        final Intent intent = getIntent();
+        if (intent == null) {
+            finish();
+            return;
+        }
+        switch (intent.getAction()) {
+            case MessagingPushConstants.NotificationAction.OPENED:
+                handlePushOpen(intent);
+                break;
+            case MessagingPushConstants.NotificationAction.BUTTON_CLICKED:
+                handlePushButtonClicked(intent);
+                break;
+            case MessagingPushConstants.NotificationAction.DISMISSED:
+                handlePushDismiss(intent);
+                break;
+            default:
+                break;
+        }
+        finish();
+    }
+
+
+    /**
+     * Handles the push notification open action.
+     *
+     * @param intent the intent received from the push notification interaction
+     */
+    private void handlePushOpen(final Intent intent) {
+        Messaging.handleNotificationResponse(intent, true, null);
+        executePushAction(intent);
+    }
+
+    /**
+     * Handles clicks on push notification custom buttons.
+     *
+     * @param intent the intent received from interacting with buttons on push notification
+     */
+    private void handlePushButtonClicked(final Intent intent) {
+        final String actionId = intent.getStringExtra(MessagingPushConstants.Tracking.Keys.ACTION_ID);
+        Messaging.handleNotificationResponse(intent, true, actionId);
+
+        // Dismiss the notification once interacted
+        final String messageId = intent.getStringExtra(MessagingPushConstants.Tracking.Keys.MESSAGE_ID);
```

Comment:
same with message id, should this be null checked first?

Reply:
Yes thanks for the feedback.. We are using messageId.hashCode().. It must be null checked to avoid NPE